src/algorithms/superviewport.ts

```diff
@@ -0,0 +1,142 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AbstractViewportAlgorithm,
+  AlgorithmInput,
+  AlgorithmOutput,
+} from "./core";
+import SuperCluster, { ClusterFeature } from "supercluster";
+import { MarkerUtils, Marker } from "../marker-utils";
+import { Cluster } from "../cluster";
+import { getPaddedViewport } from "./utils";
+import equal from "fast-deep-equal";
+
+export type SuperClusterViewportOptions = SuperCluster.Options<
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  { [name: string]: any },
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  { [name: string]: any }
+> & { viewportPadding?: number };
+
+export interface SuperClusterViewportState {
+  zoom: number;
+  view: [number, number, number, number];
+}
+
+/**
+ * A very fast JavaScript algorithm for geospatial point clustering using KD trees.
+ *
+ * @see https://www.npmjs.com/package/supercluster for more information on options.
+ */
+export class SuperClusterViewportAlgorithm extends AbstractViewportAlgorithm {
+  protected superCluster: SuperCluster;
+  protected markers: Marker[];
+  protected clusters: Cluster[];
+  protected state: SuperClusterViewportState;
+
+  constructor({
+    maxZoom,
+    radius = 60,
+    viewportPadding = 60,
+    ...options
+  }: SuperClusterViewportOptions) {
+    super({ maxZoom, viewportPadding });
+
+    this.superCluster = new SuperCluster({
+      maxZoom: this.maxZoom,
+      radius,
+      ...options,
+    });
+
+    this.state = { zoom: 0, view: [0, 0, 0, 0] };
+  }
+
+  public calculate({
+    map,
+    mapCanvasProjection,
+    markers,
+  }: AlgorithmInput): AlgorithmOutput {
+    const state: SuperClusterViewportState = {
+      zoom: Math.round(map.getZoom()),
+      view: getPaddedViewport(
+        map.getBounds(),
+        mapCanvasProjection,
+        this.viewportPadding
+      ),
+    };
+
+    let changed = !equal(this.state, state);
+    if (!equal(markers, this.markers)) {
+      changed = true;
+      // TODO use proxy to avoid copy?
+      this.markers = [...markers];
+
+      const points = this.markers.map((marker) => {
+        const position = MarkerUtils.getPosition(marker);
+        const coordinates = [position.lng(), position.lat()];
+        return {
+          type: "Feature" as const,
+          geometry: {
+            type: "Point" as const,
+            coordinates,
+          },
+          properties: { marker },
+        };
+      });
+      this.superCluster.load(points);
+    }
+
+    this.state = state;
+
+    if (changed) {
+      this.clusters = this.cluster();
+    }
+
+    return { clusters: this.clusters, changed };
+  }
+
+  public cluster(): Cluster[] {
```

Comment:
cluster should return the un-cached version (that is you should not use the state here but derived the current value from the map)

Reply:
Fixed